citations-impl/impl/src/java/org/sakaiproject/citation/impl/soloapi/AbstractConverter.java

```diff
@@ -0,0 +1,96 @@
+package org.sakaiproject.citation.impl.soloapi;
+
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.node.ArrayNode;
+import org.sakaiproject.citation.api.Citation;
+import org.sakaiproject.citation.api.CitationService;
+import org.sakaiproject.citation.api.Context;
+import org.sakaiproject.citation.api.Converter;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.ArrayList;
+import java.util.Vector;
+
+public abstract class AbstractConverter implements Converter {
+
+	private CitationService citationService;
+
+	protected JsonNode primoNMBibNode;
+	protected JsonNode docNode;
+	protected JsonNode getItNode;
+
+	public void setCitationService(CitationService citationService) {
+		this.citationService = citationService;
+	}
+
+	protected void setCitationProperty(Citation citation, JsonNode primoNMBibNode, String citationKey, ArrayList<String> jsonPath) {
+
+		// find the json node
+		for (String jsonNode : jsonPath) {
+			primoNMBibNode = primoNMBibNode.get(jsonNode);
+			if (primoNMBibNode==null) {
+				break;
+			}
+		}
+
+		if (primoNMBibNode!=null){
+			String value = null;
+
+			// if jsonnode contains array
+			if (primoNMBibNode instanceof ArrayNode){
+				for (JsonNode jsonNode : primoNMBibNode) {
+					value = jsonNode.getTextValue();
+					if (value != null) {
+						citation.setCitationProperty(citationKey, value);
+					}
+				}
+			}
+			else {
+				value = primoNMBibNode.getTextValue();
+				if (value != null) {
+					value = formatSpecialCases(citationKey, value);
+					citation.setCitationProperty(citationKey, value);
+				}
+			}
+
+		}
+	}
+
+	private String formatSpecialCases(String citationKey, String value) {
+		if (citationKey.equals("year") && value.length()>3) {
+			value = value.substring(0,4);
+		}
+		else if (citationKey.equals("otherIds") && value.startsWith("http://oxfordsfx")) {
+			for (String string : value.split("&")) {
+				if (string.startsWith("rft_id") && !string.startsWith("rft_id=info:")){
```

Comment:
This line is also supposed to fix a bug where some of the open url we are getting from the bodleian have multiple rft_id paramters